File: 백준/Silver I/1074. Z/Z.py
import sys
input=sys.stdin.readline

def z(x, i, j, ans):
    if x==1:
        if i-2==r and j-2==c:
            print(ans-3)
        elif i-2==r and j-1==c:
            print(ans-2)
        elif i-1==r and j-2==c:
            print(ans-1)
        elif i-1==r and j-1==c:
            print(ans)
    else:
        nx=2**(x-1)
        nnx=2**(2*(x-1))
        if r<i-nx and c<j-nx:
            z(x-1, i-nx, j-nx, ans-3*nnx)
        elif r<i-nx and j-nx<=c:
            z(x-1, i-nx, j, ans-2*nnx)
        elif i-nx<=r and c<j-nx:
            z(x-1, i, j-nx, ans-nnx)
        else:
            z(x-1, i, j, ans)

n, r, c = map(int, input().split())
z(n, 2**n, 2**n, 2**(2*n)-1)

File: 백준/Silver/1074. Z/Z.java
/**
 * @author 이지원
 * @date 24.02.14
 * @link https://www.acmicpc.net/problem/1074
 * @keyword_solution 
 * 재귀
 * 배열을 4구역으로 나누고, 주어진 좌표가 어느 구역에 있는지 재귀적으로 탐색
 * @input 배열의 크기 N, 좌표 r, c
 * @output r행 c열을 몇 번째로 방문했는지 출력
 * @time_complex 
 * @perf 
 */

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	static StringBuilder sb = new StringBuilder();
	static int n, r, c, N, ans;

	public static void main(String[] args) throws IOException {
		st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		r = Integer.parseInt(st.nextToken());
		c = Integer.parseInt(st.nextToken());
		N = (int) Math.pow(2, n);
		z(n, 0, 0);
		System.out.println(ans);
	}

	static void z(int d, int x, int y) {
		if (d == 0) {
			return;
		} else {
			int idx = (int) Math.pow(2, d - 1);
			int tmp = idx * idx;
			int nx = x + idx, ny = y + idx;
			if (r < nx && c < ny) {
				z(d - 1, x, y);
			} else if (r < nx && ny <= c) {
				ans += tmp;
				z(d - 1, x, ny);
			} else if (nx <= r && c < ny) {
				ans += 2 * tmp;
				z(d - 1, nx, y);
			} else {
				ans += 3 * tmp;
				z(d - 1, nx, ny);
			}
		}
	}
}